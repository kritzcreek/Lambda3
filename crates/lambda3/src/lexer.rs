use logos::Logos;
use num_derive::{FromPrimitive, ToPrimitive};
use smol_str::SmolStr;

pub fn lex_str(input: &str) -> Vec<Item> {
    let mut tokens = vec![];
    let lexer = Lexer::new(input);

    for item in lexer {
        if (item.1).0 == SyntaxKind::EOF {
            tokens.push(item);
            return tokens;
        } else {
            tokens.push(item);
        }
    }
    unreachable!()
}

pub struct Lexer<'a> {
    inner: logos::Lexer<'a, SyntaxKind>,
}

impl<'a> Lexer<'a> {
    pub fn new(input: &'a str) -> Self {
        Self {
            inner: SyntaxKind::lexer(input),
        }
    }
}

fn is_whitespace(kind: SyntaxKind) -> bool {
    kind == SyntaxKind::WHITESPACE
}

pub type Token = (SyntaxKind, SmolStr);
pub type Item = (Vec<Token>, Token);

impl<'a> Iterator for Lexer<'a> {
    type Item = (Vec<Token>, Token);

    fn next(&mut self) -> Option<Self::Item> {
        let mut leading = vec![];

        loop {
            let kind = self.inner.next();
            let text = self.inner.slice();

            match kind {
                None => return Some((leading, (SyntaxKind::EOF, "".into()))),
                Some(kind) => {
                    if !is_whitespace(kind) {
                        return Some((leading, (kind, text.into())));
                    } else {
                        leading.push((kind, text.into()));
                    }
                }
            }
        }
    }
}

#[derive(Debug, Copy, Clone, PartialEq, Logos, FromPrimitive, ToPrimitive)]
#[repr(u16)]
#[allow(bad_style, missing_docs, unreachable_pub)]
pub enum SyntaxKind {
    #[regex(r"\s")]
    WHITESPACE,

    #[token("let")]
    LET_KW,

    #[token("true")]
    TRUE_KW,

    #[token("false")]
    FALSE_KW,

    #[token("Bool")]
    BOOL,

    #[token("Int")]
    INT,

    #[regex("[A-Za-z][A-Za-z0-9]*")]
    IDENT,

    #[regex("[0-9]+")]
    NUMBER_LIT,

    #[token(".")]
    DOT,

    #[token("+")]
    PLUS,

    #[token("-")]
    MINUS,

    #[token("*")]
    STAR,

    #[token("/")]
    SLASH,

    #[token("\\")]
    BACKSLASH,

    #[token("=")]
    EQUALS,

    #[token("(")]
    L_PAREN,

    #[token(")")]
    R_PAREN,

    #[token("[")]
    L_BRACKET,

    #[token("]")]
    R_BRACKET,

    #[token("{")]
    L_BRACE,

    #[token("}")]
    R_BRACE,

    #[token("<")]
    LT,

    #[token(">")]
    GT,

    #[token("?")]
    QUESTION,

    #[token(":")]
    COLON,

    #[token(";")]
    SEMICOLON,

    #[token(",")]
    COMMA,

    #[token("->")]
    ARROW,

    #[token("_")]
    UNDERSCORE,

    #[error]
    LEXER_ERROR,
    ERROR,

    EOF,

    // Types
    PAREN_TY,
    INT_TY,
    BOOL_TY,
    FUNC_TY,

    TY_ARG,
    TY_RES,

    // Expressions
    PAREN_E,       // a parenthesized expression
    LITERAL_E,     // a literal
    VAR_E,         // wraps a WORD token
    LAMBDA_E,      // a func expression
    APPLICATION_E, // a function application

    ROOT,

    // Patterns
    PAREN_P,
    VAR_P,
    WILDCARD_P,
    ANNOTATION_P,
}

#[macro_export]
macro_rules ! T {
    [;] => { SyntaxKind::SEMICOLON };
    [,] => { SyntaxKind::COMMA };
    ['('] => { SyntaxKind::L_PAREN };
    [')'] => { SyntaxKind::R_PAREN };
    ['{'] => { SyntaxKind::L_BRACE };
    ['}'] => { SyntaxKind::R_BRACE };
    ['['] => { SyntaxKind::L_BRACKET };
    [']'] => { SyntaxKind::R_BRACKET };
    [<] => { SyntaxKind::L_ANGLE };
    [>] => { SyntaxKind::R_ANGLE };
    [backslash] => { SyntaxKind::BACKSLASH};
    [#] => { SyntaxKind::HASH };
    [?] => { SyntaxKind::QUESTION };
    [+] => { SyntaxKind::PLUS };
    [*] => { SyntaxKind::STAR };
    [/] => { SyntaxKind::SLASH };
    [^] => { SyntaxKind::CARET };
    [%] => { SyntaxKind::PERCENT };
    [_] => { SyntaxKind::UNDERSCORE };
    [.] => { SyntaxKind::DOT };
    [:] => { SyntaxKind::COLON };
    [=] => { SyntaxKind::EQUALS };
    [==] => { SyntaxKind::DOUBLE_EQUALS };
    [!] => { SyntaxKind::BANG };
    [-] => { SyntaxKind::MINUS };
    [->] => { SyntaxKind::ARROW };
    [<:] => { SyntaxKind::SUB };
    [actor] => { SyntaxKind::ACTOR_KW };
    [class] => { SyntaxKind::CLASS_KW };
    [object] => { SyntaxKind::OBJECT_KW };
    [async] => { SyntaxKind::ASYNC_KW };
    [true] => { SyntaxKind::TRUE_KW };
    [false] => { SyntaxKind::FALSE_KW };
    [null] => { SyntaxKind::NULL_KW };
    [func] => { SyntaxKind::FUNC_KW };
    [for] => { SyntaxKind::FOR_KW };
    [if] => { SyntaxKind::IF_KW };
    [else] => { SyntaxKind::ELSE_KW };
    [let] => { SyntaxKind::LET_KW };
    [switch] => { SyntaxKind::SWITCH_KW };
    [module] => { SyntaxKind::MODULE_KW };
    [var] => { SyntaxKind::VAR_KW };
    [prim] => { SyntaxKind::PRIM_KW };
    [query] => { SyntaxKind::QUERY_KW };
    [shared] => { SyntaxKind::SHARED_KW };
    [return] => { SyntaxKind::RETURN_KW };
    [true] => { SyntaxKind::TRUE_KW };
    [try] => { SyntaxKind::TRY_KW };
    [type] => { SyntaxKind::TYPE_KW };
    [while] => { SyntaxKind::WHILE_KW };
    [number_lit] => { SyntaxKind::NUMBER_LIT };
    [ident] => { SyntaxKind::IDENT };
    [Int] => { SyntaxKind::INT };
    [Bool] => { SyntaxKind::BOOL };
}
