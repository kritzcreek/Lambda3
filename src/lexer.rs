use logos::Logos;
use num_derive::{FromPrimitive, ToPrimitive};
use rowan::SmolStr;

pub fn lex_str<'a>(input: &'a str) -> Vec<Item> {
    let mut tokens = vec![];
    let mut lexer = Lexer::new(input);

    while let Some(item) = lexer.next() {
        if item.1.0 == SyntaxKind::Eof {
            tokens.push(item);
            return tokens
        } else {
            tokens.push(item);
        }
    }
    unreachable!()
}

pub struct Lexer<'a> {
    inner: logos::Lexer<'a, SyntaxKind>,
    lookahead: Option<(SyntaxKind, &'a str)>,
}

impl<'a> Lexer<'a> {
    pub fn new(input: &'a str) -> Self {
        Self {
            inner: SyntaxKind::lexer(input),
            lookahead: None,
        }
    }
}

fn is_whitespace(kind: SyntaxKind) -> bool {
    kind == SyntaxKind::Whitespace
}

pub type Item = (
    Vec<(SyntaxKind, SmolStr)>,
    (SyntaxKind, SmolStr),
);

impl<'a> Iterator for Lexer<'a> {
    type Item = (
        Vec<(SyntaxKind, SmolStr)>,
        (SyntaxKind, SmolStr),
    );

    fn next(&mut self) -> Option<Self::Item> {
        let mut leading = vec![];

        loop {
            let kind = self.inner.next();
            let text = self.inner.slice();

            match kind {
                None => return Some((leading, (SyntaxKind::Eof, "".into()))),
                Some(kind) => {
                    if !is_whitespace(kind) {
                        return Some((leading, (kind, text.into())))
                    }
                    else {
                        leading.push((kind, text.into()));
                    }
                }
            }
        }
    }
}

#[derive(Debug, Copy, Clone, PartialEq, Logos, FromPrimitive, ToPrimitive)]
#[repr(u16)]
pub enum SyntaxKind {
    #[regex(r"\s")]
    Whitespace,

    #[token("let")]
    LetKw,

    #[token("true")]
    TrueKw,

    #[token("false")]
    FalseKw,

    #[token("Bool")]
    Bool,

    #[token("Int")]
    Int,

    #[regex("[A-Za-z][A-Za-z0-9]*")]
    Ident,

    #[regex("[0-9]+")]
    NumberLit,

    #[token(".")]
    Dot,

    #[token("+")]
    Plus,

    #[token("-")]
    Minus,

    #[token("*")]
    Star,

    #[token("/")]
    Slash,

    #[token("\\")]
    Backslash,

    #[token("=")]
    Equals,

    #[token("(")]
    LParen,

    #[token(")")]
    RParen,

    #[token("[")]
    LBracket,

    #[token("]")]
    RBracket,

    #[token("{")]
    LBrace,

    #[token("}")]
    RBrace,

    #[token("<")]
    Lt,

    #[token(">")]
    Gt,

    #[token("?")]
    Question,

    #[token(":")]
    Colon,

    #[token(";")]
    Semicolon,

    #[token(",")]
    Comma,

    #[token("->")]
    Arrow,

    #[token("_")]
    Underscore,

    #[error]
    LexerError,
    Error,

    Eof,

    // Types
    ParenTy,
    IntTy,
    BoolTy,
    FuncTy,

    // Expressions
    ParenE, // a parenthesized expression
    LiteralE,     // a literal
    VarE,         // wraps a WORD token
    LambdaE,        // a func expression
    ApplicationE, // a function application

    Root,

    // Patterns
    ParenP,
    VarP,
    WildcardP,
    AnnotationP,
}
