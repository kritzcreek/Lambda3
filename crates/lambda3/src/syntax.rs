use crate::lexer::SyntaxKind;
use num_traits::{FromPrimitive, ToPrimitive};
pub(crate) mod ast;
#[allow(unused)]
#[allow(non_snake_case)]
#[allow(clippy::all)]
pub mod nodes;

/// Second, implementing the `Language` trait teaches rowan to convert between
/// these two SyntaxKind types, allowing for a nicer SyntaxNode API where
/// "kinds" are values from our `enum SyntaxKind`, instead of plain u16 values.
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Lang {}

pub type SyntaxNode = rowan::SyntaxNode<Lang>;
pub type SyntaxToken = rowan::SyntaxToken<Lang>;
#[allow(unused)]
pub type SyntaxElement = rowan::NodeOrToken<SyntaxNode, SyntaxToken>;
pub type SyntaxNodeChildren = rowan::SyntaxNodeChildren<Lang>;
#[allow(unused)]
pub type SyntaxElementChildren = rowan::SyntaxElementChildren<Lang>;

impl rowan::Language for Lang {
    type Kind = SyntaxKind;

    fn kind_from_raw(raw: rowan::SyntaxKind) -> Self::Kind {
        Self::Kind::from_u16(raw.0).unwrap()
    }

    fn kind_to_raw(kind: Self::Kind) -> rowan::SyntaxKind {
        rowan::SyntaxKind(kind.to_u16().unwrap())
    }
}

// Manually implemented accessors
pub enum LiteralKind {
    Int(i32, SyntaxToken),
    Bool(bool, SyntaxToken),
}

impl nodes::LiteralE {
    pub fn kind(&self) -> Option<LiteralKind> {
        if let Some(num_token) = self.number_lit_token() {
            Some(LiteralKind::Int(
                str::parse::<i32>(num_token.text()).unwrap(),
                num_token,
            ))
        } else if let Some(true_token) = self.true_token() {
            Some(LiteralKind::Bool(true, true_token))
        } else if let Some(false_token) = self.false_token() {
            Some(LiteralKind::Bool(false, false_token))
        } else {
            None
        }
    }
}

impl nodes::VarP {
    pub fn name(&self) -> String {
        self.ident_token().unwrap().text().to_string()
    }
}

impl nodes::VarE {
    pub fn name(&self) -> String {
        self.ident_token().unwrap().text().to_string()
    }
}
