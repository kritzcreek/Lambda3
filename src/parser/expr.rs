use super::*;
use crate::lexer::SyntaxKind::*;

/// The outcome of parsing a single S-expression
pub enum ExprRes {
    /// An expression was successfully parsed
    Ok,
    /// An unexpected token was found
    Lul(String),
}

pub fn expr(p: &mut Parser) -> ExprRes {
    let mut is_application = false;
    let checkpoint = p.checkpoint();
    match atom(p) {
        None => return ExprRes::Lul("Expected expression".to_string()),
        Some(ExprRes::Ok) => (),
        Some(ExprRes::Lul(err)) => {
            p.errors.push(err.clone());
            return ExprRes::Lul(err);
        }
    }
    loop {
        match atom(p) {
            None => break,
            Some(ExprRes::Lul(s)) => {
                p.report_error(s);
            }
            Some(ExprRes::Ok) => {
                p.start_node_at(checkpoint, ApplicationE);
                is_application = true;
            }
        }
    }
    if is_application {
        p.finish_node();
    }
    ExprRes::Ok
}

// Option<ExprRes>
// Some(OK) => atom, continue
// Some(Err) => handle recovery
// None => no atom, no tokens

fn atom(p: &mut Parser) -> Option<ExprRes> {
    match p.current() {
        LParen => parse_parenthesized(p),
        Ident => parse_var(p),
        NumberLit | TrueKw | FalseKw => parse_literal(p),
        Backslash => return parse_lambda(p),
        LexerError => p.bump_any(),
        _ => return None,
    }

    Some(ExprRes::Ok)
}

fn parse_parenthesized(p: &mut Parser) {
    p.start_node(ParenE);
    p.bump(LParen);
    expr(p);

    if !p.eat(RParen) {
        p.start_node(Error);
        p.errors.push(format!(
            "unexpected token {:?}, expected ')'",
            p.current()
        ));

        while p.current() != RParen && p.current() != Eof {
            p.bump_any()
        }

        p.finish_node();
        p.eat(RParen);
    }

    p.finish_node();
}

fn parse_var(p: &mut Parser) {
    p.start_node(VarE);
    p.bump(Ident);
    p.finish_node();
}

fn parse_literal(p: &mut Parser) {
    p.start_node(LiteralE);
    p.bump_any();
    p.finish_node();
}

fn parse_lambda(p: &mut Parser) -> Option<ExprRes> {
    p.start_node(LambdaE);
    p.bump(Backslash);

    let checkpoint = p.checkpoint();
    if p.eat(Ident) {
        p.start_node_at(checkpoint, VarP);
        p.finish_node();
    } else {
        p.report_error(format!("expected binder, got {:?}", p.current()))
    }

    if p.eat(Colon) {
        types::typ(p)
    }

    if !p.eat(Dot) {
        p.finish_node();
        return Some(ExprRes::Lul(format!(
            "expected '.', got {:?}",
            p.current()
        )));
    }

    match expr(p) {
        ExprRes::Ok => p.finish_node(),
        ExprRes::Lul(err) => {
            p.report_error(err);
            p.finish_node();
            return None;
        }
    }

    Some(ExprRes::Ok)
}

