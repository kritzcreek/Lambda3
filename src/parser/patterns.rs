use crate::lexer::SyntaxKind::*;
use super::*;

pub fn parse_pattern(p: &mut Parser) {
    let checkpoint = p.checkpoint();

    let current = p.current();
    match current {
        Ident => {
            p.bump(Ident);
            p.finish_at(checkpoint, VarP);
        }
        Underscore => {
            p.bump(Underscore);
            p.finish_at(checkpoint, WildcardP);
        }
        LParen => {
            p.bump(LParen);
            parse_pattern(p);
            if !p.eat(RParen) {
                p.report_error(format!("missing closing paren, got {:?}", p.current()));
            }
            p.finish_at(checkpoint, ParenP);
        }
        _ => {
            p.report_error(format!("expected pattern, got {:?}", current));
            return;
        }
    }

    opt_annotation(p, checkpoint);
}

fn opt_annotation(p: &mut Parser, checkpoint: Checkpoint) -> bool {
    if p.eat(Colon) {
        types::typ(p);
        p.finish_at(checkpoint, AnnotationP);
        return true;
    }

    return false;
}