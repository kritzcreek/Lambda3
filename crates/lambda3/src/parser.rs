use rowan::{Checkpoint, GreenNode, GreenNodeBuilder, Language};

use crate::cst;
use crate::lexer::SyntaxKind::{EOF, ERROR, ROOT};
use crate::lexer::{lex_str, SyntaxKind, Token};
use crate::syntax::ast::AstNode;
use crate::syntax::nodes;
use crate::syntax::{Lang, SyntaxNode};

mod expressions;
mod patterns;
mod types;

/// The parse results are stored as a "green tree".
/// We'll discuss working with the results later
#[derive(Debug)]
pub struct Parse {
    pub green_node: GreenNode,
    #[allow(unused)]
    pub errors: Vec<String>,
}

pub struct Parser {
    /// input tokens, including whitespace,
    /// in *reverse* order.
    tokens: Vec<(Vec<Token>, Token)>,
    /// the in-progress tree.
    builder: GreenNodeBuilder<'static>,
    /// the list of syntax errors we've accumulated
    /// so far.
    errors: Vec<String>,
}

impl Parse {
    pub fn syntax(&self) -> SyntaxNode {
        SyntaxNode::new_root(self.green_node.clone())
    }

    pub fn expr(&self) -> nodes::Expr {
        self.syntax()
            .children()
            .find_map(|n| nodes::Expr::cast(n))
            .unwrap()
    }

    pub fn debug_tree(&self) -> String {
        let syntax_node = SyntaxNode::new_root(self.green_node.clone());
        let formatted = format!("{:#?}", syntax_node);

        // We cut off the last byte because formatting the SyntaxNode adds on a newline at the end.
        formatted[0..formatted.len() - 1].to_string()
    }
}

impl Parser {
    fn new(tokens: Vec<(Vec<Token>, Token)>) -> Self {
        Parser {
            tokens,
            builder: GreenNodeBuilder::new(),
            errors: Vec::new(),
        }
    }

    fn start_node(&mut self, kind: SyntaxKind) {
        self.builder.start_node(Lang::kind_to_raw(kind));
    }

    fn start_node_at(&mut self, checkpoint: Checkpoint, kind: SyntaxKind) {
        self.builder
            .start_node_at(checkpoint, Lang::kind_to_raw(kind));
    }

    fn finish_node(&mut self) {
        self.builder.finish_node();
    }

    fn finish_at(&mut self, c: Checkpoint, kind: SyntaxKind) {
        self.start_node_at(c, kind);
        self.finish_node();
    }

    fn checkpoint(&self) -> Checkpoint {
        self.builder.checkpoint()
    }

    fn report_error(&mut self, msg: String) {
        self.start_node(ERROR);
        self.errors.push(msg);
        self.finish_node();
    }

    fn nth(&self, n: usize) -> SyntaxKind {
        let len = self.tokens.len();
        if n >= len {
            SyntaxKind::EOF
        } else {
            self.tokens[len - n - 1].1 .0
        }
    }

    /// Peek at the first unprocessed token
    fn current(&self) -> SyntaxKind {
        self.nth(0)
    }

    fn at(&self, kind: SyntaxKind) -> bool {
        self.current() == kind
    }

    fn eat(&mut self, kind: SyntaxKind) -> bool {
        if !self.at(kind) {
            return false;
        }
        self.bump_any();
        true
    }

    /// Advance one token, adding it to the current branch of the tree builder.
    fn bump_any(&mut self) {
        if self.at(EOF) {
            return;
        }
        let (leading, (kind, text)) = self.tokens.pop().unwrap();

        for (kind, text) in leading {
            self.builder.token(Lang::kind_to_raw(kind), text.as_str());
        }

        self.builder.token(Lang::kind_to_raw(kind), text.as_str());
    }

    fn bump(&mut self, kind: SyntaxKind) {
        assert!(self.eat(kind))
    }
}

pub fn parse(text: &str) -> Parse {
    let mut tokens = lex_str(text);
    tokens.reverse();
    let mut p = Parser::new(tokens);

    p.start_node(ROOT);
    let _ = expressions::expr(&mut p);
    if !p.at(SyntaxKind::EOF) {
        p.start_node(ERROR);
        while !p.at(EOF) {
            p.bump_any()
        }
        p.finish_node();
        p.errors.push(format!("Unexpected token {:?}", p.current()))
    }
    p.finish_node();

    Parse {
        green_node: p.builder.finish(),
        errors: p.errors,
    }
}

pub fn parse_type(text: &str) -> Parse {
    let mut tokens = lex_str(text);
    tokens.reverse();
    let mut p = Parser::new(tokens);

    p.start_node(ROOT);
    let _ = types::typ(&mut p);
    if !p.at(SyntaxKind::EOF) {
        p.start_node(ERROR);
        while !p.at(EOF) {
            p.bump_any()
        }
        p.finish_node();
        p.errors.push(format!("Unexpected token {:?}", p.current()))
    }
    p.finish_node();

    Parse {
        green_node: p.builder.finish(),
        errors: p.errors,
    }
}

pub fn parse_pattern(text: &str) -> Parse {
    let mut tokens = lex_str(text);
    tokens.reverse();
    let mut p = Parser::new(tokens);

    p.start_node(ROOT);
    let _ = patterns::parse_pattern(&mut p);
    if !p.at(SyntaxKind::EOF) {
        p.start_node(ERROR);
        while !p.at(EOF) {
            p.bump_any()
        }
        p.finish_node();
        p.errors.push(format!("Unexpected token {:?}", p.current()))
    }
    p.finish_node();

    Parse {
        green_node: p.builder.finish(),
        errors: p.errors,
    }
}
