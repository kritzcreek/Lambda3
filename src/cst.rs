use num_traits::{FromPrimitive, ToPrimitive};
/// You can construct GreenNodes by hand, but a builder
/// is helpful for top-down parsers: it maintains a stack
/// of currently in-progress nodes
use rowan::GreenNodeBuilder;
use rowan::SmolStr;
/// GreenNode is an immutable tree, which is cheap to change,
/// but doesn't contain offsets and parent pointers.
use rowan::{Checkpoint, GreenNode};

use crate::lexer::{lex_str, SyntaxKind};
use crate::parser::{self, Parse, Parser};
use crate::syntax::{Lang, SyntaxNode, SyntaxToken};

pub fn parse(text: &str) -> Parse {
    parser::parse(text)
}

pub fn parse_type(text: &str) -> Parse {
    parser::parse_type(text)
}

pub fn parse_pattern(text: &str) -> Parse {
    parser::parse_pattern(text)
}

macro_rules! ast_node {
    ($ast:ident, $kind:ident) => {
        #[derive(PartialEq, Eq, Hash, Debug)]
        #[repr(transparent)]
        pub struct $ast(pub SyntaxNode);
        impl $ast {
            #[allow(unused)]
            pub fn cast(node: SyntaxNode) -> Option<Self> {
                if node.kind() == SyntaxKind::$kind {
                    Some(Self(node))
                } else {
                    None
                }
            }
        }
    };
}

macro_rules! lit_node {
    ($ast:ident, $kind:ident) => {
        #[derive(PartialEq, Eq, Hash, Debug)]
        #[repr(transparent)]
        pub struct $ast(pub SyntaxNode);
        impl $ast {
            #[allow(unused)]
            fn cast(node: SyntaxNode) -> Option<Self> {
                if node.kind() == SyntaxKind::LiteralE
                    && first_token(&node)?.kind() == SyntaxKind::$kind
                {
                    Some(Self(node))
                } else {
                    None
                }
            }
        }
    };
    ($ast:ident, $kind1:ident, $kind2:ident) => {
        #[derive(PartialEq, Eq, Hash, Debug)]
        #[repr(transparent)]
        pub struct $ast(pub SyntaxNode);
        impl $ast {
            #[allow(unused)]
            fn cast(node: SyntaxNode) -> Option<Self> {
                if node.kind() == SyntaxKind::LiteralE {
                    let token = first_token(&node)?.kind();
                    if token == SyntaxKind::$kind1 || token == SyntaxKind::$kind2 {
                        Some(Self(node))
                    } else {
                        None
                    }
                } else {
                    None
                }
            }
        }
    };
}

ast_node!(Root, Root);
ast_node!(Var, VarE);
ast_node!(Lambda, LambdaE);
ast_node!(Application, ApplicationE);
lit_node!(IntLit, NumberLit);
lit_node!(BooleanLit, TrueKw, FalseKw);

#[derive(PartialEq, Eq, Hash, Debug)]
#[repr(transparent)]
pub struct Expr(SyntaxNode);

#[derive(Debug)]
pub enum ExprKind {
    Var(Var),
    Lambda(Lambda),
    Application(Application),
    IntLit(IntLit),
    BooleanLit(BooleanLit),
}

impl Expr {
    fn cast(node: SyntaxNode) -> Option<Self> {
        if Var::cast(node.clone()).is_some()
            || Lambda::cast(node.clone()).is_some()
            || Application::cast(node.clone()).is_some()
            || IntLit::cast(node.clone()).is_some()
            || BooleanLit::cast(node.clone()).is_some()
        {
            Some(Expr(node))
        } else {
            None
        }
    }

    pub fn kind(&self) -> ExprKind {
        Var::cast(self.0.clone())
            .map(ExprKind::Var)
            .or_else(|| Lambda::cast(self.0.clone()).map(ExprKind::Lambda))
            .or_else(|| Application::cast(self.0.clone()).map(ExprKind::Application))
            .or_else(|| IntLit::cast(self.0.clone()).map(ExprKind::IntLit))
            .or_else(|| BooleanLit::cast(self.0.clone()).map(ExprKind::BooleanLit))
            .unwrap()
    }
}

impl Root {
    pub fn expr(&self) -> Option<Expr> {
        self.0.children().find_map(Expr::cast)
    }
}

pub fn first_token(node: &SyntaxNode) -> Option<SyntaxToken> {
    node.children_with_tokens().find_map(|node| {
        if node.kind() != SyntaxKind::Whitespace {
            node.as_token().cloned()
        } else {
            None
        }
    })
}

pub fn first_word(node: &SyntaxNode) -> Option<String> {
    node.children_with_tokens().find_map(|node| {
        if node.kind() == SyntaxKind::Ident {
            Some(node.as_token().unwrap().text().to_string())
        } else {
            None
        }
    })
}

impl Lambda {
    pub fn binder(&self) -> Option<String> {
        self.0
            .children()
            .find(|node| node.kind() == SyntaxKind::VarP)
            .and_then(|node| first_word(&node))
    }

    pub fn body(&self) -> Option<Expr> {
        self.0.children().find_map(Expr::cast)
    }
}

impl Var {
    pub fn name(&self) -> String {
        first_word(&self.0).unwrap()
    }
}

impl IntLit {
    pub fn value(&self) -> i32 {
        first_token(&self.0).unwrap().text().parse().unwrap()
    }
}

impl BooleanLit {
    pub fn value(&self) -> bool {
        first_token(&self.0).unwrap().kind() == SyntaxKind::TrueKw
    }
}

impl Application {
    pub fn func(&self) -> Expr {
        self.0.children().filter_map(Expr::cast).nth(0).unwrap()
    }

    pub fn arg(&self) -> Expr {
        self.0.children().filter_map(Expr::cast).nth(1).unwrap()
    }
}

#[cfg(test)]
mod tests {
    use insta::{assert_debug_snapshot, glob};

    use super::*;

    #[test]
    fn parse_expr_test() {
        use std::fs;

        glob!("examples/expr/*.l3", |path| {
            let input = fs::read_to_string(path).unwrap();
            let parse = parse(&input);
            let node = parse.syntax();

            assert_debug_snapshot!((input, parse.errors, node));
        });
    }

    #[test]
    fn parse_type_test() {
        use std::fs;

        glob!("examples/types/*.l3", |path| {
            let input = fs::read_to_string(path).unwrap();
            let parse = parse_type(&input);
            let node = parse.syntax();

            assert_debug_snapshot!((input, parse.errors, node));
        });
    }

    #[test]
    fn parse_patterns_test() {
        use std::fs;

        glob!("examples/patterns/*.l3", |path| {
            let input = fs::read_to_string(path).unwrap();
            let parse = parse_pattern(&input);
            let node = parse.syntax();

            assert_debug_snapshot!((input, parse.errors, node));
        });
    }
}
