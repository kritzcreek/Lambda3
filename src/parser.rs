use rowan::{Checkpoint, GreenNode, GreenNodeBuilder, Language, SmolStr};

use crate::cst::Root;
use crate::lexer::SyntaxKind::{Eof, Error, RootE, Whitespace};
use crate::lexer::{lex_str, SyntaxKind};
use crate::syntax::{Lang, SyntaxNode};

mod expr;

/// The parse results are stored as a "green tree".
/// We'll discuss working with the results later
#[derive(Debug)]
pub struct Parse {
    pub green_node: GreenNode,
    #[allow(unused)]
    pub errors: Vec<String>,
}

pub struct Parser {
    /// input tokens, including whitespace,
    /// in *reverse* order.
    tokens: Vec<(Vec<(SyntaxKind, SmolStr)>, (SyntaxKind, SmolStr))>,
    /// the in-progress tree.
    builder: GreenNodeBuilder<'static>,
    /// the list of syntax errors we've accumulated
    /// so far.
    errors: Vec<String>,
}

impl Parse {
    pub fn syntax(&self) -> SyntaxNode {
        SyntaxNode::new_root(self.green_node.clone())
    }

    pub fn root(&self) -> Root {
        Root::cast(self.syntax()).unwrap()
    }

    pub fn debug_tree(&self) -> String {
        let syntax_node = SyntaxNode::new_root(self.green_node.clone());
        let formatted = format!("{:#?}", syntax_node);

        // We cut off the last byte because formatting the SyntaxNode adds on a newline at the end.
        formatted[0..formatted.len() - 1].to_string()
    }
}

impl Parser {
    fn new(tokens: Vec<(Vec<(SyntaxKind, SmolStr)>, (SyntaxKind, SmolStr))>) -> Self {
        Parser {
            tokens,
            builder: GreenNodeBuilder::new(),
            errors: Vec::new(),
        }
    }

    fn start_node(&mut self, kind: SyntaxKind) {
        self.builder.start_node(Lang::kind_to_raw(kind));
    }

    fn start_node_at(&mut self, checkpoint: Checkpoint, kind: SyntaxKind) {
        self.builder
            .start_node_at(checkpoint, Lang::kind_to_raw(kind));
    }

    fn finish_node(&mut self) {
        self.builder.finish_node();
    }

    fn finish_at(&mut self, c: Checkpoint, kind: SyntaxKind) {
        self.start_node_at(c, kind);
        self.finish_node();
    }

    fn checkpoint(&self) -> Checkpoint {
        self.builder.checkpoint()
    }

    fn report_error(&mut self, msg: String) {
        self.start_node(Error);
        self.errors.push(msg);
        self.finish_node();
    }

    fn nth(&self, n: usize) -> SyntaxKind {
        let len = self.tokens.len();
        if n >= len {
            SyntaxKind::Eof
        } else {
            self.tokens[len - n - 1].1 .0
        }
    }

    /// Peek at the first unprocessed token
    fn current(&self) -> SyntaxKind {
        self.nth(0)
    }

    fn nth_at(&self, n: usize, kind: SyntaxKind) -> bool {
        self.nth(n) == kind
    }

    fn at(&self, kind: SyntaxKind) -> bool {
        self.current() == kind
    }

    fn eat(&mut self, kind: SyntaxKind) -> bool {
        if self.current() != kind {
            return false;
        }
        self.bump_any();
        true
    }

    /// Advance one token, adding it to the current branch of the tree builder.
    fn bump_any(&mut self) {
        if self.current() == Eof {
            return;
        }
        let (leading, (kind, text)) = self.tokens.pop().unwrap();

        for (kind, text) in leading {
            self.builder.token(Lang::kind_to_raw(kind), text);
        }

        self.builder.token(Lang::kind_to_raw(kind), text);
    }

    fn bump(&mut self, kind: SyntaxKind) {
        assert!(self.eat(kind))
    }
}

pub fn parse(text: &str) -> Parse {
    let mut tokens = lex_str(text);
    tokens.reverse();
    let mut p = Parser::new(tokens);

    p.start_node(RootE);
    let _ = expr::expr(&mut p);
    if p.current() != SyntaxKind::Eof {
        p.start_node(Error);
        while p.current() != Eof {
            p.bump_any()
        }
        p.finish_node();
        p.errors.push(format!("Unexpected token {:?}", p.current()))
    }
    p.finish_node();

    Parse {
        green_node: p.builder.finish(),
        errors: p.errors.clone(),
    }
}
