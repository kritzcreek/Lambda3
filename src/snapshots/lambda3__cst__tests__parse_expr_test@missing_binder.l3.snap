---
source: src/cst.rs
expression: "(input, parse.errors, node)"
input_file: src/examples/expr/missing_binder.l3
---
(
    "(\\. x)",
    [
        "expected pattern, got Dot",
    ],
    Root@0..6
      ParenE@0..6
        LParen@0..1 "("
        LambdaE@1..5
          Backslash@1..2 "\\"
          Error@2..2
          Dot@2..3 "."
          VarE@3..5
            Whitespace@3..4 " "
            Ident@4..5 "x"
        RParen@5..6 ")"
    ,
)
